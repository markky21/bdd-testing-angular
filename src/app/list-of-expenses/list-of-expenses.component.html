<mat-card class="Wrapper">
  <mat-card-header>
    <mat-card-title>List of expenses</mat-card-title>
  </mat-card-header>
  <mat-card-content>
    <app-month-selector></app-month-selector>
    <app-expense-table></app-expense-table>
    <app-expense-form></app-expense-form>
  </mat-card-content>
</mat-card>
