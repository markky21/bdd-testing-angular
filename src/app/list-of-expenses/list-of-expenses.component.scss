.Form {
  margin: 32px;
}
