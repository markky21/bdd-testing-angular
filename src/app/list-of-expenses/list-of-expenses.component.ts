import { Component } from '@angular/core';

@Component({
  selector: 'app-list-of-expenses',
  templateUrl: './list-of-expenses.component.html',
  styleUrls: ['./list-of-expenses.component.scss'],
})
export class ListOfExpensesComponent {}
