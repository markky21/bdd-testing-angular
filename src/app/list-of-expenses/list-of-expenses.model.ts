export enum ExpenseCategory {
  HOME = 'HOME',
  CAR = 'CAR',
}

export interface Expense {
  id?: number;
  date: string;
  amount: number;
  category: ExpenseCategory;
}
