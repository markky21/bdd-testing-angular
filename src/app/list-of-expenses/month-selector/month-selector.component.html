<mat-form-field appearance="fill" class="Wrapper">
  <mat-label>Month and Year</mat-label>
  <input [attr.test-id]="'inputMonthSelector'" [max]="maxDate" matInput [matDatepicker]="dp" [formControl]="date" />
  <mat-datepicker-toggle [attr.test-id]="'inputMonthSelectorToggle'" matSuffix [for]="dp"></mat-datepicker-toggle>
  <mat-datepicker
    #dp
    startView="year"
    (yearSelected)="chosenYearHandler($event)"
    (monthSelected)="chosenMonthHandler($event, dp)"
  >
  </mat-datepicker>
</mat-form-field>
