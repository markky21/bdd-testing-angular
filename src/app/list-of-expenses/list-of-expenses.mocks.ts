import * as moment from 'moment';
import { Expense, ExpenseCategory } from './list-of-expenses.model';

export const mockExpenseList: Expense[] = [
  {
    id: 1,
    date: moment().add(-5, 'days').toISOString(),
    amount: 15.5,
    category: ExpenseCategory.CAR,
  },
  {
    id: 2,
    date: moment().add(-9, 'days').toISOString(),
    amount: 163.5,
    category: ExpenseCategory.HOME,
  },
  {
    id: 3,
    date: moment().add(-10, 'days').toISOString(),
    amount: 5.5,
    category: ExpenseCategory.HOME,
  },
];

export const getMockExpenseList: () => Expense[] = () => [
  {
    id: 1,
    date: moment().add(-5, 'days').toISOString(),
    amount: 15.5,
    category: ExpenseCategory.CAR,
  },
  {
    id: 2,
    date: moment().add(-9, 'days').toISOString(),
    amount: 163.5,
    category: ExpenseCategory.HOME,
  },
  {
    id: 3,
    date: moment().add(-10, 'days').toISOString(),
    amount: 5.5,
    category: ExpenseCategory.HOME,
  },
];
