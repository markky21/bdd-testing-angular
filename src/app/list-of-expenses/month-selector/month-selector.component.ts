import { Component, OnDestroy, OnInit } from '@angular/core';
import { FormControl } from '@angular/forms';
import { ListOfExpensesService } from '../list-of-expenses.service';
import { DateAdapter, MAT_DATE_FORMATS, MAT_DATE_LOCALE } from '@angular/material/core';
import { DATE_FORMATS } from '../../utils/dates.utils';
import { MAT_MOMENT_DATE_ADAPTER_OPTIONS, MomentDateAdapter } from '@angular/material-moment-adapter';
import * as moment from 'moment';
import { MatDatepicker } from '@angular/material/datepicker';
import { Subject } from 'rxjs';
import { takeUntil, tap } from 'rxjs/operators';

@Component({
  selector: 'app-month-selector',
  templateUrl: './month-selector.component.html',
  styleUrls: ['./month-selector.component.scss'],
  providers: [
    {
      provide: DateAdapter,
      useClass: MomentDateAdapter,
      deps: [MAT_DATE_LOCALE, MAT_MOMENT_DATE_ADAPTER_OPTIONS],
    },
    { provide: MAT_DATE_FORMATS, useValue: DATE_FORMATS },
  ],
})
export class MonthSelectorComponent implements OnInit, OnDestroy {
  private destroyed$ = new Subject();
  date = new FormControl(moment());
  maxDate = new Date();

  constructor(public listOfExpensesService: ListOfExpensesService) {}

  ngOnInit(): void {
    this.listOfExpensesService.selectedMonth$.next(this.date.value);

    this.date.valueChanges
      .pipe(
        tap((value) => this.listOfExpensesService.selectedMonth$.next(value)),
        takeUntil(this.destroyed$)
      )
      .subscribe();
  }

  ngOnDestroy(): void {}

  chosenYearHandler(normalizedYear: moment.Moment): void {
    const ctrlValue = this.date.value;
    ctrlValue.year(normalizedYear.year());
    this.date.setValue(ctrlValue);
  }

  chosenMonthHandler(normalizedMonth: moment.Moment, datepicker: MatDatepicker<moment.Moment>): void {
    const ctrlValue = this.date.value;
    ctrlValue.month(normalizedMonth.month());
    this.date.setValue(ctrlValue);
    datepicker.close();
  }
}
