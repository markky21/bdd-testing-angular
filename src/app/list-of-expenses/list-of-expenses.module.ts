import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ListOfExpensesComponent } from './list-of-expenses.component';
import { MatCardModule } from '@angular/material/card';
import { MatButtonModule } from '@angular/material/button';
import { ExpenseTableComponent } from './expense-table/expense-table.component';
import { MonthSelectorComponent } from './month-selector/month-selector.component';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { MatDatepickerModule } from '@angular/material/datepicker';
import { MatNativeDateModule } from '@angular/material/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatTableModule } from '@angular/material/table';
import { ExpenseFormComponent } from './expense-form/expense-form.component';
import {MatSelectModule} from '@angular/material/select';

@NgModule({
  declarations: [ListOfExpensesComponent, ExpenseTableComponent, MonthSelectorComponent, ExpenseFormComponent],
  imports: [
    CommonModule,
    MatCardModule,
    MatButtonModule,
    MatFormFieldModule,
    MatInputModule,
    MatDatepickerModule,
    MatNativeDateModule,
    ReactiveFormsModule,
    MatTableModule,
    MatSelectModule,
  ],
  exports: [ListOfExpensesComponent],
})
export class ListOfExpensesModule {}
