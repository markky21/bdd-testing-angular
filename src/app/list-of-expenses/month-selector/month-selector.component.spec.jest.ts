import {ComponentFixture, TestBed} from '@angular/core/testing';
import {MonthSelectorComponent} from './month-selector.component';
import {ListOfExpensesService} from '../list-of-expenses.service';
import MockDate from 'mockdate';
import {MatInputModule} from '@angular/material/input';
import {MatDatepickerModule} from '@angular/material/datepicker';
import {ReactiveFormsModule} from '@angular/forms';
import {MatNativeDateModule} from '@angular/material/core';
import {NoopAnimationsModule} from '@angular/platform-browser/animations';
import moment from 'moment';
import {DATE_FORMATS} from '../../utils/dates.utils';
import {Overlay, OverlayContainer} from '@angular/cdk/overlay';
import {ListOfExpensesServiceStub} from '../list-of-expenses.service.stub';

describe('MonthSelectorComponent', () => {
  const currentDate = new Date(2020, 3, 21);
  let component: MonthSelectorComponent;
  let fixture: ComponentFixture<MonthSelectorComponent>;
  let listOfExpensesService: ListOfExpensesService;
  let overlay: Overlay;
  let overlayContainer: OverlayContainer;

  const getTemplate = (): HTMLElement => fixture.debugElement.nativeElement;
  const getInputMothSelector = () =>
    getTemplate().querySelector('input[test-id="inputMonthSelector"]') as HTMLInputElement | null;
  const getInputMothSelectorToggle = () =>
    getTemplate().querySelector('[test-id="inputMonthSelectorToggle"] button') as HTMLButtonElement | null;
  const getMatCalendarContent = () =>
    overlayContainer.getContainerElement().querySelector('mat-calendar') as HTMLElement | null;

  beforeAll(() => {
    MockDate.set(currentDate);
  });

  afterAll(() => {
    MockDate.reset();
  });

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [MonthSelectorComponent],
      providers: [{ provide: ListOfExpensesService, useClass: ListOfExpensesServiceStub }],
      imports: [MatInputModule, MatDatepickerModule, ReactiveFormsModule, MatNativeDateModule, NoopAnimationsModule],
    }).compileComponents();

    listOfExpensesService = TestBed.inject(ListOfExpensesService);
    overlayContainer = TestBed.inject(OverlayContainer);
    overlay = TestBed.inject(Overlay);

    fixture = TestBed.createComponent(MonthSelectorComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should display page components', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should initially set to current month', () => {
    fixture.detectChanges();
    const spySelectedMonths$ = spyOn(listOfExpensesService.selectedMonth$, 'next');
    component.ngOnInit();
    expect(spySelectedMonths$).toHaveBeenCalledWith(moment());
    expect(getInputMothSelector()?.value).toContain(moment().format(DATE_FORMATS.display.dateInput).toString());
  });

  it('should on destroy clean subscription', () => {
    const spyListOfExpensesService = spyOn(listOfExpensesService.selectedMonth$, 'next');

    component.ngOnDestroy();
    // tslint:disable-next-line:no-non-null-assertion
    getInputMothSelector()!.value = moment().add(-1, 'months').format(DATE_FORMATS.display.dateInput).toString();
    getInputMothSelector()?.dispatchEvent(new Event('input'));

    // Check if new value was send
    expect(spyListOfExpensesService).not.toHaveBeenCalled();
  });

  it('should allow user to select months in the past', async () => {
    const spyListOfExpensesService = spyOn(listOfExpensesService.selectedMonth$, 'next');
    getInputMothSelectorToggle()?.click();
    fixture.detectChanges();
    expect(getMatCalendarContent()).toBeTruthy();

    // Select previous month in datepicker
    getMatCalendarContent()
      ?.querySelectorAll('.mat-calendar-body .mat-calendar-body-cell')
      .item(2)
      .dispatchEvent(new MouseEvent('click'));
    fixture.detectChanges();
    expect(getInputMothSelector()?.value).toContain(
      moment(new Date(2020, 2, 21)).format(DATE_FORMATS.display.dateInput).toString()
    );

    // Check if new value was send
    expect(spyListOfExpensesService).toHaveBeenCalledWith(moment().add(-1, 'months'));
  });

  it('should disallow user to select months in the feature', () => {
    getInputMothSelectorToggle()?.click();
    fixture.detectChanges();
    expect(getMatCalendarContent()).toBeTruthy();

    // Select month in future in datepicker
    getMatCalendarContent()
      ?.querySelectorAll('.mat-calendar-body .mat-calendar-body-cell')
      .item(4)
      .dispatchEvent(new MouseEvent('click'));
    fixture.detectChanges();
    expect(getInputMothSelector()?.value).toContain(moment().format(DATE_FORMATS.display.dateInput).toString());
  });

  it('should allow user to select year in the past', async () => {
    const spyListOfExpensesService = spyOn(listOfExpensesService.selectedMonth$, 'next');
    getInputMothSelectorToggle()?.click();
    fixture.detectChanges();
    expect(getMatCalendarContent()).toBeTruthy();

    // Select previous year in datepicker
    getMatCalendarContent()?.querySelector('.mat-calendar-period-button')?.dispatchEvent(new MouseEvent('click'));
    fixture.detectChanges();
    getMatCalendarContent()?.querySelector('.mat-calendar-period-button')?.dispatchEvent(new MouseEvent('click'));
    fixture.detectChanges();
    getMatCalendarContent()
      ?.querySelectorAll('.mat-calendar-body-cell')
      ?.item(0)
      .dispatchEvent(new MouseEvent('click'));
    fixture.detectChanges();

    // Select previous month in datepicker
    getMatCalendarContent()
      ?.querySelectorAll('.mat-calendar-body .mat-calendar-body-cell')
      .item(2)
      .dispatchEvent(new MouseEvent('click'));
    fixture.detectChanges();
    expect(getInputMothSelector()?.value).toContain(
      moment(new Date(1997, 2, 21)).format(DATE_FORMATS.display.dateInput).toString()
    );

    // Check if new value was send
    expect(spyListOfExpensesService.calls.argsFor(0).toString()).toEqual(moment(new Date(1997, 2, 21)).toString());
  });
});
