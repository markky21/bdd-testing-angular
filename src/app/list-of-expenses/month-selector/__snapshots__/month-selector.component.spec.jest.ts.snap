// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthSelectorComponent should display page components 1`] = `
<app-month-selector
  date={[Function FormControl]}
  destroyed$={[Function Subject]}
  listOfExpensesService={[Function ListOfExpensesServiceStub]}
  maxDate={[Function Date]}
>
  <mat-form-field
    appearance="fill"
    class="Wrapper mat-form-field mat-primary mat-form-field-type-mat-input mat-form-field-appearance-fill mat-form-field-can-float mat-form-field-should-float mat-form-field-has-label ng-untouched ng-pristine ng-valid _mat-animation-noopable"
  >
    <div
      class="mat-form-field-wrapper"
    >
      <div
        class="mat-form-field-flex"
      >
        
        
        <div
          class="mat-form-field-infix"
        >
          <input
            aria-haspopup="dialog"
            aria-invalid="false"
            aria-required="false"
            class="mat-datepicker-input mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid cdk-text-field-autofill-monitored"
            data-mat-calendar="mat-datepicker-0"
            matinput=""
            max="2020-04-21T00:00:00+02:00"
            test-id="inputMonthSelector"
          />
          <mat-datepicker
            startview="year"
          />
          <span
            class="mat-form-field-label-wrapper"
          >
            
            <label
              class="mat-form-field-label"
            >
              
              
              <mat-label>
                Month and Year
              </mat-label>
              
              
            </label>
          </span>
        </div>
        
        <div
          class="mat-form-field-suffix"
        >
          <mat-datepicker-toggle
            class="mat-datepicker-toggle"
            data-mat-calendar="mat-datepicker-0"
            matsuffix=""
            test-id="inputMonthSelectorToggle"
          >
            <button
              aria-haspopup="dialog"
              aria-label="Open calendar"
              class="mat-focus-indicator mat-icon-button mat-button-base _mat-animation-noopable"
              mat-icon-button=""
              tabindex="0"
              type="button"
            >
              <span
                class="mat-button-wrapper"
              >
                
                <svg
                  class="mat-datepicker-toggle-default-icon"
                  fill="currentColor"
                  focusable="false"
                  height="24px"
                  viewBox="0 0 24 24"
                  width="24px"
                >
                  <path
                    d="M19 3h-1V1h-2v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V8h14v11zM7 10h5v5H7z"
                  />
                </svg>
              </span>
              <span
                class="mat-button-ripple mat-ripple mat-button-ripple-round"
                matripple=""
              />
              <span
                class="mat-button-focus-overlay"
              />
            </button>
          </mat-datepicker-toggle>
        </div>
      </div>
      
      <div
        class="mat-form-field-underline"
      >
        <span
          class="mat-form-field-ripple"
        />
      </div>
      <div
        class="mat-form-field-subscript-wrapper"
      >
        
        
        <div
          class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
          style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
        >
          
          <div
            class="mat-form-field-hint-spacer"
          />
        </div>
      </div>
    </div>
  </mat-form-field>
</app-month-selector>
`;
